### PR TITLE
sot-190-add_job-worker-handler

### PR DESCRIPTION
**Pull request by Draply**
# Check List

- [x] Don't forget to squash commits into meaningful chunks before merging
- [x] Check every test passed.
- [x] Did you split imports into std, third parties, and custom parts?
- [x] Format your code.
  - [x] cargo fmt
  - [x] cargo clippy
  - [x] `taplo fmt --config taplo/taplo.toml`
- [x] Check unwrap, expect().
- [x] Check clone().
- [x] Check your commit messages.
- [x] Have you added meaningful comments.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -51,20 +50,26 @@ scopeguard = { version = "1.2.0" }
...
-stone-cli = { git = "https://github.com/sota-zk-labs/stone-cli" }
+stone-cli = { git = "https://github.com/zksecurity/stone-cli.git", branch = "main"}
 tempfile = { version = "3.13.0" }
 test-log = { version = "0.2.16" }
-testcontainers-modules = { version = "0.11.2" }
+testcontainers-modules = { version = "0.11.3" }
 thiserror = { version = "1.0.64" }
 tokio = { version = "1.39.3", features = ["macros", "rt-multi-thread"] }
 tonic = { version = "0.12.2" }
+toml = "0.8.19"
```
Comment (zk-steve) [inline Cargo.toml]: version

File: crates/worker/src/main.rs [hunk 2]
```diff
@@ -21,7 +33,7 @@ use crate::options::Options;
 #[tokio::main]
 async fn main() {
     let options: Options = CliArgs::default_run_or_get_options(env!("APP_VERSION"));
-
+    println!("options: {:?}", options);
```
Comment (zk-steve) [inline crates/worker/src/main.rs]: ?

File: crates/worker/src/jobs/verify_proof_and_register_job.rs [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+
+use crate::state::State;
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct VerifyAndRegisterJob(Worker<PayloadVerifyJob>);
+
+impl TaskHandler for VerifyAndRegisterJob {
+    const IDENTIFIER: &'static str = VERIFY_PROOF_AND_REGISTER_IDENTIFIER;
+    #[instrument(level = "info", skip(self, _ctx))]
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_proof_and_register_job.rs]:
```suggestion

    #[instrument(level = "info", skip(self, _ctx))]
```

File: crates/worker/src/jobs/register_continuous_page_job.rs [hunk 4]
```diff
@@ -0,0 +1,44 @@
+use crate::state::State;
+use graphile_worker::{IntoTaskHandlerResult, TaskHandler, WorkerContext};
+use irelia_adapter::aptos_writer::contracts_caller::memory_page_fact_registry::extract_register_memory::extract_register_continuous_page;
+use irelia_adapter::aptos_writer::contracts_caller::memory_page_fact_registry::register_continuous_memory_page::register_continuous_memory_page;
+use irelia_common::workers::{Worker, REGISTER_CONTINUOUS_IDENTIFIER};
+use irelia_core::entities::payload_verify_job::PayloadVerifyJob;
+use serde::{Deserialize, Serialize};
+use tracing::log::info;
+use tracing::{instrument, Span};
+use tracing_opentelemetry::OpenTelemetrySpanExt;
```
Comment (zk-steve) [inline crates/worker/src/jobs/register_continuous_page_job.rs]: format

File: crates/worker/Cargo.toml [hunk 5]
```diff
@@ -5,28 +5,36 @@ name = "irelia_worker"
...
+    "macros",
+    "runtime-tokio",
 ], default-features = false }
 testcontainers-modules = { workspace = true, features = ["postgres", "redis"] }
 thiserror = { workspace = true }
 tokio = { workspace = true, features = ["full"] }
 tower-http = { workspace = true, features = ["timeout", "trace"] }
 tracing = { workspace = true }
 tracing-opentelemetry = { workspace = true }
+uuid = { version = "1.10.0", features = ["v4"] }
```
Comment (zk-steve) [inline crates/worker/Cargo.toml]: workspace = true

File: crates/adapter/src/aptos_writer/flow_test.rs [hunk 6]
```diff
@@ -49,7 +49,7 @@ mod tests {
                 let now = Instant::now();
                 aptos_container
                     .upload_contract(
-                        "../contracts/navori",
+                        "/home/ubuntu/code/irelia/contracts",
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/flow_test.rs]: ?

File: crates/adapter/Cargo.toml [hunk 7]
```diff
@@ -13,36 +13,36 @@ aptos-testcontainer = { workspace = true }
...
-  "postgres",
-  "json",
-  "macros",
-  "runtime-tokio",
-] }
 tracing = { workspace = true }
-opentelemetry = { workspace = true }
-tracing-opentelemetry = {workspace = true}
+tracing-opentelemetry = { workspace = true }
+uuid = { version = "1.10.0", features = ["v4", "serde"] }
```
Comment (zk-steve) [inline crates/adapter/Cargo.toml]: .

File: crates/worker/src/jobs/verify_fri_job.rs [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+
+use crate::state::State;
+
+#[derive(Deserialize, Serialize, Debug)]
+pub struct VerifyFriJob(Worker<PayloadVerifyJob>);
+
+impl TaskHandler for VerifyFriJob {
+    const IDENTIFIER: &'static str = VERIFY_FRI_IDENTIFIER;
+    #[instrument(level = "info", skip(self, _ctx))]
+    async fn run(self, _ctx: WorkerContext) -> impl IntoTaskHandlerResult {
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_fri_job.rs]:
```suggestion
    async fn run(self, ctx: WorkerContext) -> impl IntoTaskHandlerResult {
```

File: crates/adapter/src/prover/stone_prover.rs [hunk 9]
```diff
@@ -34,13 +35,15 @@ const SERIALIZED_PROOF_PATH: &str = "bootloader_serialized_proof.json";
 
 pub struct StoneProver {
     pub cairo_pie: Vec<PathBuf>,
-    pub layout: LayoutName,
+    pub layout: String,
```
Comment (zk-steve) [inline crates/adapter/src/prover/stone_prover.rs:37]: Why changing from LayoutName to String?

Comment (Draply) [reply crates/adapter/src/prover/stone_prover.rs:37]: I dont want to import LayoutName to the worker,  so i convert to string and parse to layoutname

File: Cargo.toml [hunk 10]
```diff
@@ -51,20 +50,26 @@ scopeguard = { version = "1.2.0" }
...
 tracing-bunyan-formatter = { version = "0.3.9" }
 tracing-opentelemetry = { version = "0.27.0" }
 tracing-subscriber = { version = "0.3.18" }
 url = { version = "2.4.0", features = ["serde"] }
-uuid = { version = "1.10.0" }
+uuid = { version = "1.10.0", features = ["v4"] }
+reqwest = { version = "0.12.9" }
+tokio-postgres = { version = "0.7.12" }
+base64 = { version = "0.22.1" }
+zip = { version = "0.6.6" }
```
Comment (zk-steve) [inline Cargo.toml:62]: reorganize

File: Dockerfile
Comment (zk-steve) [inline Dockerfile:1]: rollback
exclude this from this PR

File: crates/adapter/src/aptos_writer/config.rs [hunk 11]
```diff
@@ -33,38 +34,37 @@ impl EnvConfig {
...
             chain_id,
             node_url,
             private_key,
-            module_address,
+            aptos_verifier_address,
         }
     }
 }
-
+#[derive(Clone)]
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/config.rs:16]:
```suggestion

#[derive(Clone)]
```

File: crates/adapter/src/aptos_writer/contracts_caller/verify_fri/extract_verify_fri_input.rs [hunk 12]
```diff
@@ -5,18 +5,15 @@ use aptos_sdk::move_types::value::MoveValue;
...
-    Vec<(
-        FriVerifyInput,
-        MoveValue,
-        MoveValue,
-        MoveValue,
-        MoveValue,
-        MoveValue,
-    )>,
-> {
+type FriVerifyInputVec = Vec<(
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/contracts_caller/verify_fri/extract_verify_fri_input.rs:8]: (
     FriVerifyInput,
     MoveValue,
     MoveValue,
     MoveValue,
     MoveValue,
     MoveValue,
 ) => to a new struct

File: crates/worker/src/state.rs
Comment (zk-steve) [inline crates/worker/src/state.rs:1]: ?

File: crates/adapter/src/prover/stone_prover.rs [hunk 13]
```diff
@@ -76,7 +79,7 @@ impl ProverPort for StoneProver {
         let proof_args = stone_cli::args::ProveBootloaderArgs {
             cairo_programs: None,
             cairo_pies: Some(self.cairo_pie.clone()),
-            layout: self.layout.clone(),
+            layout: layout.clone(),
```
Comment (zk-steve) [inline crates/adapter/src/prover/stone_prover.rs]:
```suggestion
            layout: layout,
```

File: crates/adapter/src/worker/mod.rs [hunk 14]
```diff
@@ -50,4 +54,96 @@ impl WorkerPort for WorkerAdapter {
...
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(
+                MERKLE_STATEMENT_VERIFIER_IDENTIFIER,
+                payload,
+                Default::default(),
+            )
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: map to CoreError::InternalError

File: crates/adapter/src/worker/mod.rs [hunk 15]
```diff
@@ -50,4 +54,96 @@ impl WorkerPort for WorkerAdapter {
...
+            propagator.inject_context(&cx, &mut tracing_info)
+        });
+        let payload = Worker {
+            data: &job,
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(VERIFY_FRI_IDENTIFIER, payload, Default::default())
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: .

File: crates/adapter/src/worker/mod.rs [hunk 16]
```diff
@@ -50,4 +54,96 @@ impl WorkerPort for WorkerAdapter {
...
+            propagator.inject_context(&cx, &mut tracing_info)
+        });
+        let payload = Worker {
+            data: &job,
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(REGISTER_CONTINUOUS_IDENTIFIER, payload, Default::default())
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: .

File: crates/adapter/src/worker/mod.rs [hunk 17]
```diff
@@ -50,4 +54,96 @@ impl WorkerPort for WorkerAdapter {
...
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(
+                VERIFY_PROOF_AND_REGISTER_IDENTIFIER,
+                payload,
+                Default::default(),
+            )
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: .

File: crates/worker/src/main.rs [hunk 18]
```diff
@@ -91,7 +98,27 @@ pub async fn run_workers(options: Options) {
...
+    let worker_adapter: Arc<dyn WorkerPort + Send + Sync> = Arc::new(
+        WorkerAdapter::new(
+            &options.pg.url,
+            options.pg.max_size,
+            options.worker.schema.clone(),
+        )
+        .await,
+    );
+
+    let app_config = AppConfig::from(EnvConfig::new());
```
Comment (zk-steve) [inline crates/worker/src/main.rs]: Use toml instead

File: crates/worker/src/main.rs [hunk 19]
```diff
@@ -91,7 +98,27 @@ pub async fn run_workers(options: Options) {
...
+
+    let sequence_number = app_config
+        .client
+        .get_account(app_config.account.address())
+        .await
+        .unwrap()
+        .into_inner()
+        .sequence_number;
+    app_config.account.set_sequence_number(sequence_number);
+    let state = State::new(job_port, worker_adapter, app_config);
```
Comment (zk-steve) [inline crates/worker/src/main.rs:127]: config should already inside each port, adapter,...,
```suggestion
    let state = State::new(job_port, worker_adapter);
```

File: crates/worker/src/jobs/verify_proof_and_register_job.rs [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    async fn run(self, ctx: WorkerContext) -> impl IntoTaskHandlerResult {
+        let span = Span::current();
+        let parent_cx = opentelemetry::global::get_text_map_propagator(|propagator| {
+            propagator.extract(&self.0.tracing)
+        });
+
+        span.set_parent(parent_cx);
+        let gps_verify_input = extract_gps_input(&self.0.data.sharp_proof.main_proof).unwrap();
+
+        info!("VERIFY FRI IN PROGRESS");
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_proof_and_register_job.rs]:
```suggestion
        debug!("VERIFY FRI IN PROGRESS");
```

File: crates/worker/src/jobs/verify_proof_and_register_job.rs [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+            .await
+            .unwrap();
+
+        let mut job_status = self.0.data.job;
+        job_status.status = CairoJobStatus::ONCHAIN;
+        job_status.validation_done = true;
+
+        state.job_port.update(job_status).await.unwrap();
+
+        info!("=== VERIFY ONCHAIN SUCCESS ===");
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_proof_and_register_job.rs]:
```suggestion
        debug!("=== VERIFY ONCHAIN SUCCESS ===");
```

File: crates/worker/src/jobs/verify_proof_and_register_job.rs (hunk 20 above)
Comment (zk-steve) [reply crates/worker/src/jobs/verify_proof_and_register_job.rs]: better remove this

File: crates/worker/src/jobs/verify_fri_job.rs [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+        let span = Span::current();
+        let parent_cx = opentelemetry::global::get_text_map_propagator(|propagator| {
+            propagator.extract(&self.0.tracing)
+        });
+
+        span.set_parent(parent_cx);
+        let fri_verify_inputs =
+            extract_verify_fri_input(&self.0.data.sharp_proof.fri_proofs.clone()).unwrap();
+
+        info!("VERIFY FRI IN PROGRESS");
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_fri_job.rs]: remove

File: crates/worker/src/jobs/verify_fri_job.rs [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+                &state.app_config,
+                fri_verify_input,
+                proof,
+                fri_queue,
+                evaluation_point,
+                fri_step_size,
+                expected_root.clone(),
+            )
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_fri_job.rs]:
```suggestion
            ?;
```

Comment (zk-steve) [reply crates/worker/src/jobs/verify_fri_job.rs]: Screenshot 2024-12-12 at 09 14 01

File: crates/worker/src/jobs/verify_fri_job.rs [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+            info!("Verify FRI success with expected root {:?}", expected_root);
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_fri_job.rs]: debug

File: crates/worker/src/jobs/verify_fri_job.rs (hunk 23 above)
Comment (zk-steve) [reply crates/worker/src/jobs/verify_fri_job.rs]: fix for all other jobs

File: crates/worker/src/jobs/register_continuous_page_job.rs [hunk 25]
```diff
@@ -0,0 +1,43 @@
+use graphile_worker::{IntoTaskHandlerResult, TaskHandler, WorkerContext};
+use irelia_adapter::aptos_writer::contracts_caller::memory_page_fact_registry::extract_register_memory::extract_register_continuous_page;
+use irelia_adapter::aptos_writer::contracts_caller::memory_page_fact_registry::register_continuous_memory_page::register_continuous_memory_page;
+use irelia_common::workers::{Worker, REGISTER_CONTINUOUS_IDENTIFIER};
+use irelia_core::entities::payload_verify_job::PayloadVerifyJob;
+use serde::{Deserialize, Serialize};
+use tracing::log::info;
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt;
+use crate::app_state::State;
```
Comment (zk-steve) [inline crates/worker/src/jobs/register_continuous_page_job.rs:11]: fat

Comment (zk-steve) [reply crates/worker/src/jobs/register_continuous_page_job.rs:11]: _fmt _

File: crates/worker/src/app_state.rs [hunk 26]
```diff
@@ -1,11 +1,15 @@
...
 
+use irelia_adapter::aptos_writer::config::AppConfig;
 use irelia_core::ports::job::JobPort;
+use irelia_core::ports::worker::WorkerPort;
 
 #[derive(Clone)]
 pub struct State {
     pub job_port: Arc<dyn JobPort + Send + Sync>,
+    pub worker_port: Arc<dyn WorkerPort + Send + Sync>,
+    pub app_config: AppConfig,
```
Comment (zk-steve) [inline crates/worker/src/app_state.rs:12]: remove

File: crates/worker/src/jobs/verify_proof_and_register_job.rs [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+            propagator.extract(&self.0.tracing)
+        });
+
+        span.set_parent(parent_cx);
+        let gps_verify_input = extract_gps_input(&self.0.data.sharp_proof.main_proof).unwrap();
+
+        let state = ctx.extensions().get::<State>().unwrap();
+        verify_proof_and_register(&state.app_config, &gps_verify_input)
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_proof_and_register_job.rs]:
```suggestion
            ?;
```

File: crates/worker/src/jobs/register_continuous_page_job.rs [hunk 28]
```diff
@@ -0,0 +1,41 @@
...
+        span.set_parent(parent_cx);
+        let register_continuous_page_inputs =
+            extract_register_continuous_page(&self.0.data.sharp_proof.memory_pages).unwrap();
+
+        let state = ctx.extensions().get::<State>().unwrap();
+
+        for page in register_continuous_page_inputs {
+            register_continuous_memory_page(&state.app_config, page)
+                .await
+                .unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/register_continuous_page_job.rs]:
```suggestion
                ?;
```

File: crates/worker/src/jobs/register_continuous_page_job.rs [hunk 29]
```diff
@@ -0,0 +1,41 @@
+use crate::app_state::State;
+use graphile_worker::{IntoTaskHandlerResult, TaskHandler, WorkerContext};
+use irelia_adapter::aptos_writer::contracts_caller::memory_page_fact_registry::extract_register_memory::extract_register_continuous_page;
+use irelia_adapter::aptos_writer::contracts_caller::memory_page_fact_registry::register_continuous_memory_page::register_continuous_memory_page;
+use irelia_common::workers::{Worker, REGISTER_CONTINUOUS_IDENTIFIER};
+use irelia_core::common::core_error::CoreError;
+use irelia_core::entities::payload_verify_job::PayloadVerifyJob;
+use serde::{Deserialize, Serialize};
+use tracing::Span;
+use tracing_opentelemetry::OpenTelemetrySpanExt;
```
Comment (zk-steve) [inline crates/worker/src/jobs/register_continuous_page_job.rs]: format ?

File: crates/worker/src/jobs/register_continuous_page_job.rs [hunk 30]
```diff
@@ -0,0 +1,41 @@
...
+    const IDENTIFIER: &'static str = REGISTER_CONTINUOUS_IDENTIFIER;
+    async fn run(self, ctx: WorkerContext) -> impl IntoTaskHandlerResult {
+        let span = Span::current();
+        let parent_cx = opentelemetry::global::get_text_map_propagator(|propagator| {
+            propagator.extract(&self.0.tracing)
+        });
+
+        span.set_parent(parent_cx);
+        let register_continuous_page_inputs =
+            extract_register_continuous_page(&self.0.data.sharp_proof.memory_pages).unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/register_continuous_page_job.rs]:
```suggestion
            extract_register_continuous_page(&self.0.data.sharp_proof.memory_pages)?;
```

File: crates/worker/src/jobs/merkle_statement_job.rs [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+        let verify_merkle_inputs =
+            extract_verify_merkle_input(&self.0.data.sharp_proof.merkle_proofs.clone()).unwrap();
+        for merkle_verify_input in verify_merkle_inputs {
+            debug!(
+                "Verify merkle proof with expected root: {:?}",
+                merkle_verify_input.expected_root
+            );
+            verify_merkle(&state.app_config, merkle_verify_input)
+                .await
+                .unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/merkle_statement_job.rs]:
```suggestion
                ?;
```

File: crates/worker/src/jobs/job_worker.rs [hunk 32]
```diff
@@ -0,0 +1,51 @@
...
+            .job_port
+            .get_job(self.0.data.customer_id, self.0.data.cairo_job_key)
+            .await
+            .unwrap();
+
+        let stone_prover = StoneProver {
+            layout: self.0.data.proof_layout,
+            cairo_pie: vec![PathBuf::from(self.0.data.cairo_pie)],
+        };
+        let proof = stone_prover.generate_proof().await.unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/job_worker.rs:41]:
```suggestion
        let proof = stone_prover.generate_proof().await?;
```

File: crates/adapter/src/aptos_writer/config.rs [hunk 33]
```diff
@@ -33,38 +34,38 @@ impl EnvConfig {
...
         let chain_id = get_env_var_or_panic("CHAIN_ID");
         EnvConfig {
             chain_id,
             node_url,
             private_key,
-            module_address,
+            aptos_verifier_address,
         }
     }
 }
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/config.rs:15]: remove

File: crates/adapter/src/aptos_writer/config.rs [hunk 34]
```diff
@@ -10,8 +11,8 @@ use dotenv::dotenv;
 pub struct EnvConfig {
     pub node_url: String,
     pub private_key: String,
-    pub module_address: String,
     pub chain_id: String,
+    pub aptos_verifier_address: String,
 }
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/config.rs]: remove

File: crates/adapter/src/worker/mod.rs [hunk 35]
```diff
@@ -50,4 +55,97 @@ impl WorkerPort for WorkerAdapter {
...
+        };
+
+        self.worker_utils
+            .add_raw_job(
+                MERKLE_STATEMENT_VERIFIER_IDENTIFIER,
+                payload,
+                Default::default(),
+            )
+            .await
+            .map(|_job| ())
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: ?

File: crates/adapter/src/worker/mod.rs [hunk 36]
```diff
@@ -50,4 +55,97 @@ impl WorkerPort for WorkerAdapter {
...
+            .map_err(|_err| CoreError::InternalError(_err.into()))
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]:
```suggestion
            .map_err(|err| CoreError::InternalError(err.into()))
```

File: crates/adapter/src/worker/mod.rs [hunk 37]
```diff
@@ -50,4 +55,97 @@ impl WorkerPort for WorkerAdapter {
...
+        });
+        let payload = Worker {
+            data: &job,
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(VERIFY_FRI_IDENTIFIER, payload, Default::default())
+            .await
+            .map(|_job| ())
+            .map_err(|_err| CoreError::InternalError(_err.into()))
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: .

File: crates/adapter/src/worker/mod.rs [hunk 38]
```diff
@@ -50,4 +55,97 @@ impl WorkerPort for WorkerAdapter {
...
+        });
+        let payload = Worker {
+            data: &job,
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(REGISTER_CONTINUOUS_IDENTIFIER, payload, Default::default())
+            .await
+            .map(|_job| ())
+            .map_err(|_err| CoreError::InternalError(_err.into()))
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: .

File: crates/adapter/src/worker/mod.rs [hunk 39]
```diff
@@ -50,4 +55,97 @@ impl WorkerPort for WorkerAdapter {
...
+        };
+        self.worker_utils
+            .add_raw_job(
+                VERIFY_PROOF_AND_REGISTER_IDENTIFIER,
+                payload,
+                Default::default(),
+            )
+            .await
+            .map(|_job| ())
+            .map_err(|_err| CoreError::InternalError(_err.into()))
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: .

File: crates/worker/src/jobs/job_worker.rs [hunk 40]
```diff
@@ -0,0 +1,51 @@
...
+
+        span.set_parent(parent_cx);
+
+        let state = ctx.extensions().get::<State>().unwrap();
+
+        let job = state
+            .job_port
+            .get_job(self.0.data.customer_id, self.0.data.cairo_job_key)
+            .await
+            .unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/job_worker.rs]: .

File: crates/worker/src/jobs/verify_fri_job.rs [hunk 41]
```diff
@@ -0,0 +1,55 @@
...
+    const IDENTIFIER: &'static str = VERIFY_FRI_IDENTIFIER;
+    async fn run(self, ctx: WorkerContext) -> impl IntoTaskHandlerResult {
+        let span = Span::current();
+        let parent_cx = opentelemetry::global::get_text_map_propagator(|propagator| {
+            propagator.extract(&self.0.tracing)
+        });
+
+        span.set_parent(parent_cx);
+        let fri_verify_inputs =
+            extract_verify_fri_input(&self.0.data.sharp_proof.fri_proofs.clone()).unwrap();
```
Comment (zk-steve) [inline crates/worker/src/jobs/verify_fri_job.rs]:
```suggestion
            extract_verify_fri_input(&self.0.data.sharp_proof.fri_proofs).unwrap();
```

Comment (zk-steve) [reply crates/worker/src/jobs/verify_fri_job.rs]: fix unwrap

File: crates/worker/src/main.rs [hunk 42]
```diff
@@ -91,7 +98,33 @@ pub async fn run_workers(options: Options) {
         .unwrap();
 
     let job_port = Arc::new(JobDBRepository::new(pool.clone()));
-    let state = State::new(job_port);
+
+    let app_config = AppConfig::from(EnvConfig {
```
Comment (zk-steve) [inline crates/worker/src/main.rs]: fix

File: crates/adapter/src/aptos_writer/config.rs [hunk 43]
```diff
@@ -1,70 +1,39 @@
...
 
-use anyhow::Result;
 use aptos_sdk::move_types::account_address::AccountAddress;
 use aptos_sdk::rest_client::Client;
 use aptos_sdk::types::chain_id::ChainId;
 use aptos_sdk::types::LocalAccount;
-use dotenv::dotenv;
 
-pub struct EnvConfig {
+pub struct Config {
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/config.rs:9]: Too generic!

File: crates/adapter/src/aptos_writer/config.rs [hunk 44]
```diff
@@ -1,70 +1,39 @@
...
-            chain_id,
-            node_url,
-            private_key,
-            module_address,
-        }
-    }
-}
-
+#[derive(Clone)]
 pub struct AppConfig {
```
Comment (zk-steve) [inline crates/adapter/src/aptos_writer/config.rs:17]: .

File: crates/adapter/src/worker/mod.rs [hunk 45]
```diff
@@ -50,4 +55,103 @@ impl WorkerPort for WorkerAdapter {
...
+            propagator.inject_context(&cx, &mut tracing_info)
+        });
+        let payload = Worker {
+            data: &job,
+            tracing: tracing_info,
+        };
+        self.worker_utils
+            .add_raw_job(REGISTER_CONTINUOUS_IDENTIFIER, payload, Default::default())
+            .await
+            .map(|_job| job)
```
Comment (zk-steve) [inline crates/adapter/src/worker/mod.rs]: ????????????????

**Review by zk-steve — APPROVED** [review]

**Pull request merged**